Kgn/debug profile extend vscode

karnigen opened the pull request:
The changes enhance and unify developer debugging and profiling capabilities, and they extend remote debugging for the VSCode editor."

The DEBUG file now contains information about the used editor and debugger, while the PROFILE file includes information about the used profiler.

The DEBUG file may contain the name of required editor debugger support: _vscode_, _pycharm_, _pydev_, _none_
or _vscode-script_, _pycharm-script_, _pydev-script_. Also file may contain comments or empty lines.

The PROFILE file may contain name or preferred profiler: _cprofile_, _profile_, _pyinstrument_ or _none_.

Debugging in the second case, -script variant,
- when called from inkscape then local Bash script for offline debugging .ink.sh and .ink.svg is created and script is running without debug
- when you run script .ink.sh from console, debug is enabled, or may be disabled by DEBUG file.

The change also addresses the order of Inkscape paths in sys.path by uniformly moving them to the end of the list using the PYTHONPATH variable. Inkscape uses this variable to add paths when calling Python scripts (Linux), unsure if it works the same way for Windows and Mac.

lexelby commented on inkstitch.py:
```diff
@@ -31,55 +98,95 @@
+# - in debug or profile mode we run extension or profile extension
+# - in normal mode we run extension in try/except block to catch all exceptions and hide GTK spam
+if debug_active or profile_type != "none":  # if debug or profile mode
+    print(f"Extension:'{extension_name}' Debug active:{debug_active} type:'{debug_type}' "
```
I miiight end up hating this, we'll see.  I keep my DEBUG file set most of the time.  If I don't have debug mode active in PyCharm, then it just continues on, but I still get to see warnings.  Not sure if it might be annoying to have this message every time I run an extension.

karnigen replied:
Message is just for a better initial acquaintance with the code; it disappears.

But the second part concerns me. I assume that if I have an active DEBUG file, the script switches to a waiting mode and waits for the debugger to connect and doesn't continue without the debugger.

In VSCode, I haven't found a way for the script to check if there is a debugger waiting somewhere that it could connect to. That's why, for VSCode, the script stops and waits for a connection. Perhaps PyCharm has a default port on which the debugger can communicate and check if a debugger is waiting on that port.
It's inconvenient when launched directly from Inkscape because it gets stuck until the debugger is activated.

lexelby replied:
> Message is just for a better initial acquaintance with the code; it disappears.

It doesn't disappear though.  Every time we run an Ink/Stitch extension with debug enabled, Inkscape will say "the extension produced output" and show this message, and the dev will have to dismiss the dialog.

> But the second part concerns me. I assume that if I have an active DEBUG file, the script switches to a waiting mode and waits for the debugger to connect and doesn't continue without the debugger.
> 
> In VSCode, I haven't found a way for the script to check if there is a debugger waiting somewhere that it could connect to. That's why, for VSCode, the script stops and waits for a connection. Perhaps PyCharm has a default port on which the debugger can communicate and check if a debugger is waiting on that port.
It's inconvenient when launched directly from Inkscape because it gets stuck until the debugger is activated.

Ah, yes, PyCharm (and PyDev) work the other way around: the IDE opens a listening port and waits, and the script connects when it wants to be debugged.  The pydev/pycharm code in debug.py catches and ignores connection errors.  If PyCharm isn't waiting in debug mode, Ink/Stitch doesn't wait to be debug, it just continues on as normal.

lexelby commented on lib/debug.py:
```diff
@@ -140,11 +145,45 @@ def init_debugger(self):
+                
+
         try:
-            if 'PYCHARM_REMOTE_DEBUG' in os.environ:
```
Thanks, this turned out to be kinda clunky.  I like the idea of putting the debugger type in the DEBUG file.

lexelby reviewed (approved):
Wow, this is so great.  Thank you for all your work on this!  And welcome to the project, we're really happy to have you here.  I'm glad that we'll now have vscode support, I know that's going to open the door for a whole bunch of folks.

Only one small request: could we name the files .inkstitch.sh and .inkstitch.svg?  I know it's more to type, but I don't want to accidentally confuse us with Inkscape.  Actually what do you think of `inkstitch_debug.sh` and `inkstitch_debug.svg`?  I don't see a particular reason to hide the files.

karnigen commented on inkstitch.py:
```diff
@@ -2,27 +2,94 @@
+if len(sys.argv) < 2:
+    exit(1)  # no arguments - prevent accidentally running this script
+
+prefere_pip_inkex = True  # prefer pip installed inkex over inkscape bundled inkex
```
This is just for backward compatibility with inkstitch. I understand that someone may need to test a new version of inkex library. But surely, it probably shouldn't be enabled by default. I also didn't understand why it's hardcoded there. It would certainly be better to have an option to enable it outside of the script, but where? Now there are already files like `DEBUG`, `PROFILE`. Should we add `PIP_INKEX`?

Or create a structured configuration file, for example, in **INI** format, e.g. `INKSTITCH_DEVEL.ini`, where all information for debug, profile, pip_inkex, and possibly other parameters will be stored.

kaalleen replied:
I think we really should always use the pip installed inkex rather than the inkscape delivered one for Ink/Stitch. The pip installed version is the version we will package with our releases and we want to run the very same version when we are developing Ink/Stitch. They often have updates that we want to use before the next Inkscape version is out.

karnigen commented:
> Wow, this is so great. Thank you for all your work on this! And welcome to the project, we're really happy to have you here. I'm glad that we'll now have vscode support, I know that's going to open the door for a whole bunch of folks.
> 
> One small request: could we name the files .inkstitch.sh and .inkstitch.svg? I know it's more to type, but I don't want to accidentally confuse us with Inkscape. Actually what do you think of `inkstitch_debug.sh` and `inkstitch_debug.svg`? I don't see a particular reason to hide the files.

I prefer dot files for one reason: they are listed at the top and can be easily deleted. I don't take advantage of the fact that they can be hidden, I probably don't want to hide any files during development. Using common names increases the risk of accidentally deleting important files. Or we can still use prefix '_', but these files won't be listed at the top, and they won't be as easily visible.

I prefer some common prefix so that these auxiliary or utility files are close together and easily identifiable, minimizing the risk of unintended modifications to other files. It can also be used for profile outputs.

I don't have any other practical reason for naming it one way or another. Everyone can have different preferences; it's challenging to find something that would suit everyone.

lexelby commented:
Ah, your concerns make sense.  I would note that by default, on Linux and Mac, files starting with a dot will indeed be hidden, so I think we should avoid that.

How about prefixing with "debug_", like "debug_inkstitch.py"?

lexelby commented on lib/debug_utils.py:
```diff
@@ -0,0 +1,90 @@
+# Authors: see git history
```
Any reason not to put these functions in `debug.py`?  I don't feel that file is too long, and the distinction between what would go in debug_utils.py vs debug.py isn't clear to me.

karnigen replied:
Only reason was that I needed the inkex library not to be imported yet. Not before sys.path is modified.

lexelby replied:
Oh, yup, that makes sense.  How about if we switch to local imports in the functions that need `inkex`?  Then we can have everything in `debug.py`.

karnigen replied:
I considered a local import, but changing the behavior simply by moving the import position seemed too risky to me. What if someone moves it elsewhere over time, and suddenly the program's behavior changes? Plus, there are other indirect imports that import inkex.

karnigen commented:
It seems that the differences between systems, editors, debuggers, and profilers are too significant to be covered by simple files. In the end, I merged `DEBUG` and `PROFILE` into `DEVELOP.ini,` see `DEVEL_template.ini`.  
That's why it will be possible to enable each option in the configuration file, and it's also relatively easy to add additional options.
I added environment variable INKSTITCH_OFFLINE_SCRIPT=True to signal that inkstitch is being launched offline from a bash file.
I hope I simplified it without any side effects.

lexelby commented on inkstitch.py:
```diff
@@ -2,27 +2,101 @@
+
+if len(sys.argv) < 2:
+    # no arguments - prevent accidentally running this script
+    print("No arguments given, continue without arguments?")
```
There's really never a reason to run Ink/Stitch directly without arguments.  We should exit here and give a user-friendly error in a pop-up window, because users _do_ click `inkstitch.exe` expecting it to do something.

If you don't want to do that in this PR that's fine, writing a whole wxpython app is probably more than you bargained for. :)

karnigen replied:
I just wanted to be visually neutral. I wasn't sure what library to use. In 2D, I prefer Qt, but Inkscape uses GTK, Python defaults to Tk, and Inkstitch is built on wx, which unfortunately does not support gradient lines, which is a peculiar choice for an embroidery program where gradient lines are essential. Then there's the addition of Node.js with Electron and Chromium.
I'll add a dialog in wx.

lexelby commented on inkstitch.py:
```diff
@@ -2,27 +2,101 @@
+else:
+    running_from_inkscape = True
+
+debug_active = bool((gettrace := getattr(sys, 'gettrace')) and gettrace())  # check if debugger is active on startup
```
This is the first use of `:=` in this project.  I am in favor of `:=` in general, if it improves readability, but I think in this case it kind of just complicates this line...

lexelby commented on inkstitch.py:
```diff
@@ -31,55 +105,96 @@
-if getattr(sys, 'frozen', None):
+# file DEBUG exists next to inkstitch.py - enabling debug mode depends on value of debug_type in DEBUG file
+if debug_type != 'none':
+    debug_file = ini.get("DEBUG","debug_file", fallback="debug.log")
```
Let's make an option for `debug.svg` too for consistency.  `debug.svg` is really useful for debugging graph-related things in auto-fill and the like.

karnigen replied:
I've added to DEBUG.ini
```
;;; debug file for graph related things, default: debug.svg
; debug_graph_file = debug.svg
```

lexelby commented on inkstitch.py:
```diff
@@ -31,55 +105,96 @@
 
-# ignore warnings in releases
-if getattr(sys, 'frozen', None):
+# file DEBUG exists next to inkstitch.py - enabling debug mode depends on value of debug_type in DEBUG file
```
This comment is no longer true, I think?

lexelby commented on lib/debug_utils.py:
```diff
@@ -0,0 +1,110 @@
+#        "none"                         - for no profiling
+
+
+def parse_file(filename):
```
I think this function is no longer used?

lexelby commented:
Looking good! I like the idea of pulling the config out into debug.ini.

karnigen commented:
- added wx dialog when inkstitch.exe run without parameters,
- renamed DEVEL.ini to DEBUG.ini and extending it with parameters debug_log_file and debug_svg_file,
see DEBUG_template.ini,
- added debugger 'file', to enable only file logging,
- moved all profiling into debug_utils to improve readability of inkstitch.py,
- moved sys.path manipulation to debug_utils,
- canceling shapely logging, shapely 2+ uses only exceptions instead of any io,
- removed of unnecessary function enable() from debug.py, and modifying import in inkstitch.py,
- the output directory is determined by the caller in debug and offline bash creation,
- edited comments in debug.py,

kaalleen commented on lib/debug.py:
```diff
@@ -175,7 +276,7 @@ def init_svg(self):
     def save_svg(self):
         tree = etree.ElementTree(self.svg)
-        debug_svg = os.path.join(os.path.dirname(os.path.dirname(__file__)), "debug.svg")
+        debug_svg = self.debug_dir / self.debug_svg_file
```
Thank you for your work on this so far.
Can you change this to `debug_svg = str(self.debug_dir / self.debug_svg_file)` ?
Otherwise I'll get an TypeError: File or filename expected, got 'PosixPath'

karnigen replied:
I didn't realize that as of lxml version 5.0.0, it supports the pathlib Path.

karnigen commented:
### General debugging notes:
1. to enable debugging or profiling copy `DEBUG_template.ini` to `DEBUG.ini` and edit it

#### How create bash script for offline debugging from console:
1. in `DEBUG.ini` set `create_bash_script = True`
2. call` inkstitch.py` extension from i`nkscape` to create bash script named by `bash_file_base` in `DEBUG.ini`
3. run bash script from console

#### Enable debugging:
1. set `debug_type` to one of  - vscode, pycharm, pydev, see below for details
        `debug_type = vscode`    - 'debugpy' for vscode editor 
        `debug_type = pycharm`   - 'pydevd-pycharm' for pycharm editor
        `debug_type = pydev`     - 'pydevd' for eclipse editor
2. set `debug_enable = True` in `DEBUG.ini`
    or use command line argument `-d` in bash script
    or set environment variable `INKSTITCH_DEBUG_ENABLE = True` or 1 or yes or y

#### Enable profiling:
1. set profiler_type to one of - cprofile, profile, pyinstrument
      `profiler_type = cprofile`    - 'cProfile' profiler
      `profiler_type = profile`     - 'profile' profiler
      `profiler_type = pyinstrument `- 'pyinstrument' profiler
2. set `profile_enable = True` in `DEBUG.ini`
    or use command line argument `-p` in bash script
    or set environment variable `INKSTITCH_PROFILE_ENABLE = True` or 1 or yes or y

#### Miscelaneous notes:
 - to disable debugger when running from inkscape set `disable_from_inkscape = True` in `DEBUG.ini`
 - to write debug output to file set `debug_to_file = True` in `DEBUG.ini`
 - to change various output file names see `DEBUG.ini`
 - to disable waiting for debugger to attach (vscode editor) set `wait_attach = False` in `DEBUG.ini`
 - to prefer inkscape version of inkex module over pip version set `prefer_pip_inkex = False` in `DEBUG.ini`

kaalleen commented:
Thank you for your work. I'll need to take some time to actually have a look at it. There is one further request though that I have for now... Can you open a new branch an run `make style` and correct the code style issues?
